File: TheBookLore/app/src/main/java/com/penguin/thebooklore/repository/network/RetrofitHelper.java
package com.penguin.thebooklore.repository.network;

import com.penguin.thebooklore.BuildConfig;
import com.penguin.thebooklore.repository.network.model.CollectionResponse;

import java.util.concurrent.TimeUnit;

import io.reactivex.Single;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitHelper {

    private static  RetrofitHelper retrofitHelper;

    private RetrofitHelper() {}

    public static synchronized RetrofitHelper getInstance() {
        if (retrofitHelper == null) {
            retrofitHelper = new RetrofitHelper();
        }
        return retrofitHelper;
    }

    private static IRetrofitClient getClient() {

        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.level(HttpLoggingInterceptor.Level.BODY);

        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(BuildConfig.BASE_URL)
                .client(client)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .build();


        return retrofit.create(IRetrofitClient.class);

    }

    public Single<CollectionResponse> getCollection() {
        return getClient().getCollection();
    }
}

File: TheBookLore/app/src/main/java/com/penguin/thebooklore/repository/network/RetrofitHelper.kt
package com.penguin.thebooklore.repository.network

import com.penguin.thebooklore.BuildConfig
import com.penguin.thebooklore.repository.network.model.CollectionResponse

import java.util.concurrent.TimeUnit

import io.reactivex.Single
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory

object RetrofitHelper {

    val collection: Single<CollectionResponse>
        get() = client.collection

    private val client: IRetrofitClient
        get() {

            val interceptor = HttpLoggingInterceptor()
            interceptor.level = HttpLoggingInterceptor.Level.BODY

            val client = OkHttpClient.Builder()
                    .addInterceptor(interceptor)
                    .connectTimeout(30, TimeUnit.SECONDS)
                    .readTimeout(30, TimeUnit.SECONDS)
                    .build()

            val retrofit = Retrofit.Builder()
                    .baseUrl(BuildConfig.BASE_URL)
                    .client(client)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                    .build()


            return retrofit.create(IRetrofitClient::class.java)

        }

}
